### ISSUE TITLE
flycheck not automatically selecting a checker

### ISSUE THREAD

**Issue by minikN** [flycheck/flycheck#1834]
## Checklist
- [x] I have searched existing issues and StackOverflow for solutions.
- [x] I have read the [Flycheck manual][].
- [x] I have read the [Troubleshooting Guide][].
- [x] I have checked my issue has not been reported yet.

## Problem description
Hello, I'm using `lsp-mode` together with `lsp-docker` in emacs. This gives me the option to run my lsp-servers inside a docker container. For bash for example, I use `bash-language-server`. The problem with this server is however that it doesn't report any diagnostics. Because `lsp-mode` sets the default flycheck checker to `lsp` for `bash`, this results in no linting at all in `bash` or `sh` scripts.
I wanted to `shellcheck` for linting, but inside the same container to keep my host os clean. So I basically copied the checker code for `sh-shellcheck` and changed the command a bit: 

```elisp
  (flycheck-define-checker sh-shellcheck-docker
    "Use shellcheck inside a docker container"
    :command ("docker" "exec" "-i"
              (eval (concat lsp-docker-container-name "-" (number-to-string lsp-docker-container-name-suffix)))
              "shellcheck"
              "--format" "checkstyle"
              "--shell" (eval (symbol-name sh-shell))
              (option-flag "--external-sources"
                           flycheck-shellcheck-follow-sources)
              (option "--exclude" flycheck-shellcheck-excluded-warnings list
                      flycheck-option-comma-separated-list)
              "-")
    :standard-input t
    :modes sh-mode
    :error-parser flycheck-parse-checkstyle
    :error-filter (lambda (errors)
                    (flycheck-remove-error-file-names "-" errors))
    :predicate (lambda () (memq sh-shell '(bash ksh88 sh)))
    :verify
    (lambda (_)
      (let ((supported (memq sh-shell '(bash ksh88 sh))))
        (list (flycheck-verification-result-new
               :label (format "Shell %s supported" sh-shell)
               :message (if supported "yes" "no")
               :face (if supported 'success '(bold warning))))))
    :error-explainer
    (lambda (err)
      (let ((error-code (flycheck-error-id err))
            (url "https://github.com/koalaman/shellcheck/wiki/%S"))
        (and error-code `(url . ,(format url error-code))))))
```
The next step was to disable the `lsp` checker for `sh-mode` and enable my own `sh-shellcheck-docker`checker:

```elisp
  (add-hook! sh-mode
    (add-to-list 'flycheck-checkers 'sh-shellcheck-docker)
    (add-to-list 'flycheck-disabled-checkers 'lsp))
```

If I now open a `bash` or `sh` script and do `M-x flycheck-verify-setup`, this is the outoput:
```
Syntax checkers for buffer xbps-gi in sh-mode:

No checker to run in this buffer.

Checkers that could run if selected:

  sh-shellcheck-docker  select
    - may enable:           yes
    - may run:              t
    - executable:           Found at /bin/docker
    - Shell bash supported: yes

  sh-bash  select
    - may enable:    yes
    - may run:       t
    - executable:    Found at /bin/bash
    - next checkers: sh-shellcheck

Checkers that are compatible with this mode, but will not run until properly configured:

  lsp (manually disabled) enable
    - may enable: yes
    - may run:    t

  rpm-rpmlint (automatically disabled) reset
    - may enable: no
    - may run:    nil
    - executable: Not found

  sh-posix-dash
    - may enable:    yes
    - may run:       nil
    - executable:    Found at /bin/dash
    - next checkers: sh-shellcheck

  sh-posix-bash
    - may enable:    yes
    - may run:       nil
    - executable:    Found at /bin/bash
    - next checkers: sh-shellcheck

  sh-zsh (explicitly selected)
    - may enable:    yes
    - may run:       nil
    - executable:    Found at /bin/zsh
    - next checkers: sh-shellcheck

  sh-shellcheck (automatically disabled) reset
    - may enable:           no
    - may run:              t
    - executable:           Not found
    - Shell bash supported: yes

Flycheck Mode is enabled. Use SPC u C-c ! x to enable disabled checkers.

--------------------

Flycheck version: 32snapshot
Emacs version:    27.1
System:           x86_64-unknown-linux-gnu
Window system:    x
```

If I now click on `select` next to `sh-shellcheck-docker` it selects the checker and it works without any problems. Alternatively I could do `(setq flycheck-checker 'sh-shellcheck-docker)` inside my `add-hook!` to automate that. This works too.

However, shouldn't flycheck select a/my checker automatically? After all, I added it to `flycheck-checkers`.

The reason I don't want to select it manually is because if I open a `zsh` file this breaks. `shellcheck` can't be used for `zsh` but it will still be selected since it's the same mode (`sh-mode`).

So what I actually want to do is have flycheck select the appropriate checker automatically.

Thanks in advance.

## Steps to reproduce
1. Load config above
2. Open a `bash` file
3. flycheck should automatically select `sh-shellcheck-docker`
4. It does not.

- [x] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior
Automatically select appropriate checker

## System configuration
See above

Emacs configuration:
- [ ] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [x] Doom Emacs
- [ ] Other shared configuration

**Comment by JulienPalard** [comment]
Looks like flycheck is not selecting a checker automatically because lsp-diagnostics is setting `flycheck-checker` (notice the absence of `s`, it's not `flycheck-checkers`).

According to M-x describe-variable flycheck-checker:

> If set to a syntax checker only use this syntax checker and never
> select one from ‘flycheck-checkers’ automatically.

For my personal usage of using lsp for a single language, I did:

```lisp
(use-package lsp-mode
  :ensure t
  :custom
  (lsp-diagnostics-provider :none)
```

so lsp don't mess with my flycheck configuration, but you may have to do something more fine-grained.
